### DISCUSSION TITLE
Extract cell connections

### DISCUSSION THREAD

**Discussion by Licheng-Guo** [Xilinx/RapidWright#782] [Q&A]
Hey Chris,

I'm trying to extract the primitive cells connected to a given module port.

<img width="932" alt="image" src="https://github.com/Xilinx/RapidWright/assets/32432619/0e6fd97d-7222-4068-a8f8-d2ff086f01c1">

I have a basic version working by first getting all desired EDIFHierCellInsts, then looking up their physical locations. However, what if a module is encrypted and we cannot see the EDIFNetlist? Can we extract connections through the physical netlist to bypass the encryption issue? Thanks!

**Comment by clavin-xlnx** [comment]
If the design is placed, all leaf cells will be present as `Cell` objects in the `Design`.  If the design is also routed, `Net` objects with `SitePinInst` connections will be present also and can be used to trace to the `Cell` objects which have a 1:1 correspondence to their logical cell equivalents (`EDIFHierCellInst`/`EDIFCellInst`).  Tracing a `SitePinInst` back to a `Cell` pin is a bit trickier, especially if the logical view is not available, but should be doable.  There are some example helper methods in `DesignTools` that should be able to help or provide an example on how to traverse the `SiteInst` in order to see identify connections, here are a couple:
```
DesignTools.getConnnectedCells(SitePinInst)
DesignTools.getDrivingCell(SitePinInst)
```

**Reply by Licheng-Guo** [reply]
Thanks Chris, what if the design is only placed? Do you see a workaround to obtain connections between Cells?

**Reply by clavin-xlnx** [reply]
There isn't a clean solution for a placed DCP.  You could route the design quickly in Vivado and bring it back into RapidWright or you could open the DCP in Vivado and (assuming the connections are visible) you could query the net connections with Tcl commands.  But these are both slow.  

If you knew ahead of time all of the IP that were to be encrypted, you could store a cache of implementations and then keep track of just the port connections you were interested in for making connections and then rely on those cached versions.  This would require all IP to be marked with the `DONT_TOUCH` attribute in order to ensure that all versions would have the same port interface which is also undesirable.

**Reply by coherent17** [reply]
> There isn't a clean solution for a placed DCP. You could route the design quickly in Vivado and bring it back into RapidWright or you could open the DCP in Vivado and (assuming the connections are visible) you could query the net connections with Tcl commands. But these are both slow.
> 
> If you knew ahead of time all of the IP that were to be encrypted, you could store a cache of implementations and then keep track of just the port connections you were interested in for making connections and then rely on those cached versions. This would require all IP to be marked with the `DONT_TOUCH` attribute in order to ensure that all versions would have the same port interface which is also undesirable.

Hi Chris,
I'm wondering if there exist a similar method for rapidwright to return the net that connect to the cell in place and route dcp?

For example:
set net [get_nets -of_object $cells], is it possible for rapidwright to extract this connection information for further use and debug?

Thanks!
Coherent

**Reply by clavin-xlnx** [reply]
This post was originally asking about getting nets when there is encrypted cells present?  Does your design contain encrypted cells?  The problem is that RapidWright cannot look inside encrypted cells, so any net that has a source inside an encrypted cell is difficult to determine.  

In general, you can use the logical netlist to traverse connectivity of a cell and then map it back to the physical net by looking up the parent net:
```java
        Design design = Design.readCheckpoint("your_dcp.dcp");
        EDIFNetlist netlist = design.getNetlist();
        EDIFHierCellInst logicalCell = netlist.getHierCellInstFromName("your/cell/hierarchical/name");
        for (EDIFHierPortInst portInst : logicalCell.getHierPortInsts()) { 
            EDIFHierNet logicalNet = portInst.getHierarchicalNet();
            EDIFHierNet parentNet = netlist.getParentNet(logicalNet);
            Net physicalNet = design.getNet(parentNet.toString());
        }
```

**Reply by coherent17** [reply]
> This post was originally asking about getting nets when there is encrypted cells present?  Does your design contain encrypted cells?  The problem is that RapidWright cannot look inside encrypted cells, so any net that has a source inside an encrypted cell is difficult to determine.  
> 
> 
> 
> In general, you can use the logical netlist to traverse connectivity of a cell and then map it back to the physical net by looking up the parent net:
> 
> ```java
> 
>         Design design = Design.readCheckpoint("your_dcp.dcp");
> 
>         EDIFNetlist netlist = design.getNetlist();
> 
>         EDIFHierCellInst logicalCell = netlist.getHierCellInstFromName("your/cell/hierarchical/name");
> 
>         for (EDIFHierPortInst portInst : logicalCell.getHierPortInsts()) { 
> 
>             EDIFHierNet logicalNet = portInst.getHierarchicalNet();
> 
>             EDIFHierNet parentNet = netlist.getParentNet(logicalNet);
> 
>             Net physicalNet = design.getNet(parentNet.toString());
> 
>         }
> 
> ```
> 
> 
> 
> 

Hi,

Thanks, this help a lot and work great in my design!

Regards,
Coherent
